### DISCUSSION TITLE
[Bug] SkyWalkingDynamicField:_$EnhancedClassField_ws Impact on the ORM framework

### DISCUSSION THREAD

**Discussion by wl2027** [apache/skywalking#12185] [Q&A] [answered]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

Similar problem: #9357 #3629
Similar  Exception: ```MappingException: property mapping has wrong number of columns: xxx.ApmDemo._$EnhancedClassField_ws type```

If the enhanced instance object belongs to an entity class of the ORM framework, such as @Entity annotated with JPA annotations, the SkyWalkingDynamicField field needs to be declared as a non-persistent virtual field, that is, the property is not a field map of the database table.

> SkyWalkingDynamicField:_$EnhancedClassField_ws 对 ORM框架的影响
> 类似问题: #9357 #3629 
> 类似异常: ```MappingException: property mapping has wrong number of columns: xxx.ApmDemo._$EnhancedClassField_ws type```
> 
> 如果增强的实例对象属于ORM框架的实体类, 如标注了JPA注解的```@Entity```, SkyWalkingDynamicField字段需要声明为不持久的虚拟字段,即属性并非数据库表的字段映射.

![image](https://github.com/apache/skywalking/assets/152274067/4e18c62c-bcba-4a8f-ba17-5f7cb36d7acf)

### What you expected to happen

Consider to add ```transient``` SkyWalkingDynamicField keywords, or add ` ` ` @ Transient ` ` ` annotations or custom logo annotation ` ` ` @ EnhancedClassTransientField ` ` ` identified? As follows:

> 是否考虑为SkyWalkingDynamicField添加```transient```关键字,或添加```@Transient```注解或自定义标识注解```@EnhancedClassTransientField```进行标识处理?如下:

![image](https://github.com/apache/skywalking/assets/152274067/6fb8e256-b233-406e-9bd5-49dc3b6b8ab3)

### How to reproduce

Select one of the ORM framework's entity classes to enhance and then use it to manipulate the database

> 选择一个ORM框架的实体类进行增强,再使用它去操作数据库

### Are you willing to submit a pull request to fix on your own?

- [X] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wl2027** [comment]
At present, it is found that adding the keyword 'transient' 'ORM framework to ignore this field is common, and' @Transient ' 'such JPA annotations do not work for mybatis-plus, so whether to make a custom annotation' EnhancedClassTransi entField identifies which fields are skywalking enhanced, and then it is better for each ORM framework or its framework's plugins to implement the 'do not persist this field' function? As follows:

> 目前是发现直接添加关键字```transient```ORM框架对这个字段的忽略是通用,而```@Transient```这类JPA注解对mybatis-plus并不生效,所以是否做一个自定义注解```EnhancedClassTransientField```标识出来哪些是skywalking增强的字段,再由各个ORM框架或其框架的插件去实现'不持久化该字段'功能更好?如下:

![image](https://github.com/apache/skywalking/assets/152274067/84435863-f8c2-48eb-82c0-2cdb895def5b)

**Comment by wu-sheng** [comment]
I am not sure about your use case, but generally, we would not enhance Object beans, which are described in your cases.
Why do you enhance and trace such a Java to trace process?

**Reply by wl2027** [reply]
Yes, the entity objects that correspond to tables are typically simple java beans, but things like the congestion model design in DDD or the Active Record pattern in mybatis plus, there are methods that are added to the entity objects, and that happens when you need to enhance those methods

Secondly, I am not sure whether the _$EnhancedClassField_ws field added by skywalking has the need for persistence. If not, adding the keyword 'transient' can avoid similar persistence problems, and the problems caused by this kind of generated bytecode are not easy to troubleshoot To identify it can also facilitate the developer to judge the processing

>是的,一般来说表对应的实体对象是简单java bean,但是像DDD中的充血模型设计或者像mybatis-plus的Active Record模式,是有对实体对象加一些方法的,而恰好需要增强这一类方法的时候就会出现这个问题
> 
> 其次,我不清楚skywalking增加的_$EnhancedClassField_ws字段是否有持久化的需要,如果没有,加上```transient```关键字可以避免类似的持久化问题,而且这类生成字节码导致的问题并不好排查,是否加一个注解去标识一下也可以方便开发者去判断处理

**Reply by wu-sheng** [reply]
> Secondly, I am not sure whether the _$EnhancedClassField_ws field added by skywalking has the need for persistence.

It is not needed to persistent, but you didn't answer my question. Why do you enhance Java bean for ORM? We don't do that for tracing.

**Reply by wl2027** [reply]
It is not the problem found when you need to enhance the Javabeans in the orm framework, but the problem of accidentally enhancing the entity object discovery when you want to do some business tracking plug-ins, this operation should be avoided, the idea of trying to fix this vulnerability during the troubleshooting process

> 并不是需要去增强orm框架中的javabean时发现的问题,而是想做一些业务上的追踪插件时不小心增强到实体对象发现的问题,这种操作理应是可以避免的,排查过程中试图修复这个漏洞提出来的想法

**Reply by wu-sheng** [reply]
If that is your case, you need to filter things out. We can't add any annotation as that doesn't make sense to others.
You make the plugins to enhance for your private purpose, then you need to fix side effects

**Reply by wu-sheng** [reply]
Upstream only supports widely suitable cases, rather than your private ones.

**Reply by wl2027** [reply]
In addition to the above example, there are other cases, such as tracing the insert method of the mybatis Model class, which adds a dynamic attribute to the class, but when used, the entity class is directly inherited, and the field is not marked as nonpersistent, which is prone to problems

> 除了上面的例子,再有的情况有比如去追踪这个mybatis的Model类的insert方法,会给这个类加一个动态属性,但是使用的时候实体类是直接继承的,而这个字段又没有标记为非持久化,就容易有问题

![image](https://github.com/apache/skywalking/assets/152274067/f0f97420-47dd-4bab-8c36-50b15e23f84a)

**Reply by wu-sheng** [reply]
From tracing perspective, SkyWalking even doesn't need to trace ORM, JDBC driver has all we need. 
If you build a plugin for it, you need to enhance logic of ORM kernel to filter this field out.

SpringMVC plugin(official one) has a similar case, so, we provide a Spring patch to avoid the side effects of the enhancement.

**Reply by wl2027** [reply]
It is recommended to add some identification to the areas where the bytecode has changed, such as enhanced classes enhanced methods and increased fields.

Troubleshoot the problem and fix the patch is also more flexible, after all, many frameworks now have the function of modifying bytecode, and sometimes the error message in these dynamically generated bytecode is not known which framework caused

EnhanceContext, on the other hand, is only available during the load phase, and the run-time annotation to identify individuals is a better choice

> 个人建议还是给字节码变动过的地方加上一些标识吧, 如增强的类增强的方法和增加的字段 .
> 
> 排查问题的时候和或做修复补丁的时候也比较灵活,毕竟现在很多框架都有修改字节码的功能, 有时候报错信息在这些动态生成的字节码都不知道是哪个框架导致的
> 
> 而EnhanceContext这个只在加载阶段有,运行阶段注解去标识个人认为是一个较好的选择

**Reply by wu-sheng** [reply]
Everytime, adding things has risk. Same as new annotation. We have used a underlined prefix for the field name, and include ws in the name, the field name ia static. There is nothing different from an added annotation.

**Reply by wl2027** [reply]
Okay, but using an underlined prefix for field names and so on, these rules are easy to follow with other frameworks, the way you annotate them is more secure, and you can extract an sdk for developers to use.
> 好吧,不过使用带下划线的前缀作为字段名等这些规则容易和其他框架雷同,注解的方式更加保险,并且可以抽出一个sdk去给开发者使用.

Does the transient keyword on persistence still need to be modified? Need I mention pr? 
> 那么关于持久化的```transient```关键字还需要去修改吗?需要我提pr吗?

**Reply by wu-sheng** [reply]
So far, I can't see the point of having that for everyone. And I can't predict the side effects for others.
If more people and use cases show up to prove this is required, we may consider that in the future.

**Reply by wu-sheng** [reply, marked as answer]
How to filter out by an ORM extension, I believe that is not our case. We won't adopt for all frameworks, that is not reality.

As I said, if you insist to enhance Java bean for ORM, write a plugin to skip this field, this could be a supported SDK of that plugin, or another SkyWalking plugins. Both are practical.

**Reply by wl2027** [reply]
ok
